nidx/nidx_paragraph/src/request_types.rs

```diff
@@ -45,3 +46,20 @@ pub struct ParagraphSuggestRequest {
     /// Whether to do an OR/AND between prefilter results and filtering_formula
     pub filter_or: bool,
 }
+
+impl ParagraphSearchRequest {
+    // Helper method to extract valid facets from the request
+    pub fn facets(&self) -> Vec<String> {
+        self.faceted
+            .as_ref()
+            .iter()
+            .flat_map(|faceted| faceted.labels.iter())
+            .filter(|label| self.is_valid_facet(label))
+            .cloned()
+            .collect()
+    }
```

Comment:
This is converting to a `Facet` to check if it's valid and then returns the string. The return value of this is passed to `Searcher` which calls to `Facet::from_text` again, to convert to the facet needed for the query. Not sure if it's possible for this to return `Vec<Facet>` (not sure if those facets are used anywhere else).

Reply:
Now Searcher owns the `Vec<Facet>`. Unfortunately we still have to copy things around when building the response, but we do a bit less